### PR TITLE
fix find zlib.so instead of zlib.a when NDK >= 19

### PR DESCRIPTION
**Pull request by zchrissirhcz**
On Android platform, `libopencv_imgcodecs.a` is built, expected to
depend on `libz.so`. However, since Android NDK r19, NDK's `libz.a`
is found instead of `libz.so`, leading to link error
(not found libz.a) on machines without same NDK version & direcotry.

Since Android NDK-r19, toolchain pieces are installed to
`$NDK/toolchains/llvm/prebuilt/<host-tag>/...`, including `libz.so`.
Also installed to old paths (`<NDK>/platforms` and `<NDK>/sysroot`)
in NDK r19, r20, r21, but since NDK 22, old paths are removed.

- https://github.com/android/ndk/wiki/Changelog-r19
- https://github.com/android/ndk/wiki/Changelog-r22

With this commit, `libz.so` can be correctly found in NDK<19 and NDK>=19.

`ZLIB_LIBRARIES` is also simplified as `z`, by appending match (regex)
patterns for new toolchain installation directory's libz.so's paths.

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [x] I agree to contribute to the project under Apache 2 License.
- [x] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [x] The PR is proposed to proper branch
- [x] There is reference to original bug report and related work
- [x] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [x] The feature is well documented and sample code can be built with the project CMake

### CODE REVIEW

**Comment by zchrissirhcz** [comment]
Note: this commit tested locally, with cmake 3.19.4, **without** modification of `cmake-3.19.4\share\cmake-3.19\Modules\FindZLIB.cmake`.

**Comment by zchrissirhcz** [comment]
Note: this PR should only affect android platform. Android arm64 platform may fail, https://github.com/opencv/opencv/pull/19521 fix that failure and should be merged first.

**Review by alalek — COMMENTED** [review]
Thank you for contribution!

File: cmake/OpenCVFindLibsGrfmt.cmake [hunk 1]
```diff
@@ -7,9 +7,20 @@ if(BUILD_ZLIB)
...
   find_package(ZLIB "${MIN_VER_ZLIB}")
+  if(ANDROID)
+    set(CMAKE_FIND_LIBRARY_SUFFIXES ${_zlib_ORIG_CMAKE_FIND_LIBRARY_SUFFIXES})
+    unset(_zlib_ORIG_CMAKE_FIND_LIBRARY_SUFFIXES)
+  endif()
   if(ZLIB_FOUND AND ANDROID)
-    if(ZLIB_LIBRARIES MATCHES "/usr/(lib|lib32|lib64)/libz.so$")
+    if(ZLIB_LIBRARIES MATCHES "/usr/(lib|lib32|lib64)/libz.so$"
+      OR ZLIB_LIBRARIES MATCHES "/usr/lib/(aarch64-linux-android|x86_64-linux-android)/(2[1-9]|3[0-9])/libz.so"
+      OR ZLIB_LIBRARIES MATCHES "/usr/lib/(arm-linux-androideabi|i686-linux-android)/(1[6-9]|(2|3)[0-9])/libz.so"
```
[the remark this answers is not in the document]
Comment (zchrissirhcz) [reply cmake/OpenCVFindLibsGrfmt.cmake:22]: The mentioned generic pattern is simple. It's OK to use that.

For the mentioned two control variables, on android platform:
- I think there are only two origins for zlib:  build from source(libz.a), or use NDK diretories(libz.so), thus :
- `OPENCV_FORCE_ANDROID_ZLIB_Z`: is same as `NOT BUILD_ZLIB`
- `OPENCV_SKIP_ANDROID_ZLIB_Z`:  is same as `BUILD_ZLIB`

So I would like to use:
```cmake
    if(ZLIB_LIBRARIES MATCHES "/usr/lib.*/libz.so$")
      set(ZLIB_LIBRARIES z)
    endif()
```

**Review by alalek — APPROVED** [review]

**Pull request merged**